src/relay/ir/doc.cc

```diff
@@ -0,0 +1,97 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file src/tvm/relay/doc.cc
+ * \brief Doc ADT used for pretty printing.
+ * Based on Section 1 of https://homepages.inf.ed.ac.uk/wadler/papers/prettier/prettier.pdf.
+ */
+#include <memory>
+#include <vector>
+#include "doc.h"
+
+namespace tvm {
+namespace relay {
+
+// Text constructor
+DocAtom Text(const std::string& str) {
+  return std::make_shared<TextNode>(str);
+}
+
+// Line constructor
+DocAtom Line(int indent = 0) {
+  return std::make_shared<LineNode>(indent);
+}
+
+Doc::Doc(const std::string& str) {
+  if (str == "\n") {
+    this->stream_ = {Line()};
+  } else {
+    this->stream_ = {Text(str)};
+  }
+}
+
+// DSL function implementations
+
+Doc& Doc::operator<<(const Doc& right) {
+  this->stream_.insert(this->stream_.end(), right.stream_.begin(), right.stream_.end());
```

Comment:
I dont think this is correct.
Use https://en.cppreference.com/w/cpp/iterator/back_insert_iterator instead.

Reply:
Or just use a for loop.

Reply:
Also check against this == &right

Reply:
https://stackoverflow.com/questions/201718/concatenating-two-stdvectors

Reply:
You still has to catch against this == &right.

Reply:
see latest commit